File: Subarrays with K Different Integers.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var atMost = function(nums, k)
{
    const n = nums.length;
    let freq = new Map();
    let start = 0;
    let count = 0;

    for(let i = 0; i < n; i++)
    {
        freq.set(nums[i], (freq.get(nums[i]) || 0) + 1);
        while(freq.size > k)
        {
            freq.set(nums[start], freq.get(nums[start]) - 1);
            if(freq.get(nums[start]) == 0)
            {
                freq.delete(nums[start]);
            }
            start++;
        }
        count += (i - start + 1);
    }
    return count;
}
var subarraysWithKDistinct = function(nums, k) 
{
    return atMost(nums, k) - atMost(nums, k-1);
};

File: Subarrays with K Different Integers.py
class Solution(object):
    def atMost(self, nums, k):
        freq = {}
        start = count = 0
        for i in range(len(nums)):
            freq[nums[i]] = freq.get(nums[i], 0) + 1
            while len(freq) > k:
                freq[nums[start]] -= 1
                if freq[nums[start]] == 0:
                    del freq[nums[start]]
                start += 1
            count += (i - start + 1)
        return count

    def subarraysWithKDistinct(self, nums, k):
        return self.atMost(nums, k) - self.atMost(nums, k-1)

        